core/src/main/scala/kafka/log/ProducerStateManager.scala

```diff
@@ -551,21 +556,33 @@ class ProducerStateManager(val topicPartition: TopicPartition,
    */
   def oldestSnapshotOffset: Option[Long] = oldestSnapshotFile.map(file => offsetFromFilename(file.getName))
 
+  private def isProducerRetained(producerIdEntry: ProducerIdEntry, logStartOffset: Long): Boolean = {
+    producerIdEntry.lastOffset >= logStartOffset
```

Comment:
So, when logStartOffset advances, we don't update producerIdEntry.currentTxnFirstOffset, which will be used to compute lastStableOffset. Could that lead to the case that lastStableOffset is < logStartOffset?